go/service/usersearch.go

```diff
@@ -188,33 +197,159 @@ func contactSearch(mctx libkb.MetaContext, arg keybase1.UserSearchArg) (res []ke
+		phones = append(phones, phone)
+	}
+
+	lookupRes, err := provider.LookupAll(mctx, emails, phones, keybase1.RegionCode(""))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(lookupRes) > 0 {
+		var uids []keybase1.UID
+		for _, v := range lookupRes {
+			if v.Error == "" && !v.UID.IsNil() {
+				uids = append(uids, v.UID)
+			}
+		}
+
+		following, err1 := provider.FindFollowing(mctx, uids)
+		usernames, err2 := provider.FindUsernames(mctx, uids)
+		if err3 := libkb.CombineErrors(err1, err2); err3 != nil {
+			mctx.Warning("Cannot find usernames or tracking info for search results: %s", err3)
+		}
+
+		for _, v := range lookupRes {
+			// Found a resolution
+			if v.Error != "" || v.UID.IsNil() {
+				continue
+			}
+
+			assertionValue := queryString
+			if v.Coerced != "" {
+				assertionValue = v.Coerced
+			}
+			imptofu := &keybase1.ImpTofuSearchResult{
+				CoercedQuery: assertionValue,
+				Resolved:     true,
+				Uid:          v.UID,
+				Assertion:    imptofuQueryToAssertion(imptofuType, assertionValue),
+				Username:     v.UID.String(),
+			}
+			if following != nil {
+				imptofu.Following = following[v.UID]
+			}
+			if usernames != nil {
+				if uname, found := usernames[v.UID]; found {
+					imptofu.Username = uname.Username
+					imptofu.FullName = uname.Fullname
+				}
+			}
+			res = &keybase1.APIUserSearchResult{
+				Score:   1.0,
+				Imptofu: imptofu,
+			}
+			return res, nil // return here - we only want one result
+		}
+	}
+
+	// Not resolved - add SBS result.
+	imptofu := &keybase1.ImpTofuSearchResult{
+		CoercedQuery: queryString,
+		Resolved:     false,
+		Assertion:    imptofuQueryToAssertion(imptofuType, queryString),
+	}
+	res = &keybase1.APIUserSearchResult{
+		Score:   1.0,
+		Imptofu: imptofu,
+	}
+	return res, nil
+}
+
 func (h *UserSearchHandler) UserSearch(ctx context.Context, arg keybase1.UserSearchArg) (res []keybase1.APIUserSearchResult, err error) {
 	mctx := libkb.NewMetaContext(ctx, h.G()).WithLogTag("USEARCH")
 	defer mctx.TraceTimed(fmt.Sprintf("UserSearch#UserSearch(s=%q, q=%q)", arg.Service, arg.Query),
 		func() error { return err })()
 
 	if arg.Query == "" {
-		return res, nil
+		return nil, nil
+	}
+
+	if !h.disableServiceSearch {
+		res, err = doSearchRequest(mctx, arg)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	res, err = doSearchRequest(mctx, arg)
 	if arg.IncludeContacts {
 		contactsRes, err := contactSearch(mctx, arg)
 		if err != nil {
-			return nil, err
+			mctx.Error("Failed to search synced contacts: %s", err)
+		} else if len(contactsRes) > 0 {
+			res = append(contactsRes, res...)
 		}
-		if len(contactsRes) > 0 {
-			var res2 []keybase1.APIUserSearchResult
-			res2 = append(res2, contactsRes...)
-			res2 = append(res2, res...)
-			res = res2
-
-			maxRes := arg.MaxResults
-			if maxRes > 0 && len(res) > maxRes {
-				res = res[:maxRes]
+	}
+
+	if arg.ImpTofuQuery != nil {
+		imptofuRes, err := imptofuSearch(mctx, h.contactsProvider, *arg.ImpTofuQuery)
```

Comment:
is the query e164 checked for validity before burning rate limit in the gui?

Reply:
(and email regex checked)

Reply:
yes, by frontend

Reply:
although the email checker is much more powerful on serverside - even knows valid tlds - clientside check is much simpler

Reply:
hm, does that work for emails like `n@ai`? may be too strict

Reply:
is this a valid email?

clientside checker looks for `@` character and `.` that's after `@` but isn't the last char in the string.

serverside has some external validator from `validator` package